### PR TITLE
feat(components/enumeration): items with checkboxes

### PR DESCRIPTION
**Pull request by frassinier**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
Can select several items by clicking them as buttons

**What is the new behavior?**
Be able to select enumeration items by their checkboxes

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration
path for existing applications: ...

**Other information**:
http://guidelines.talend.com/document/92132#/navigation-layout/enumeration-list-maker

### CODE REVIEW

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

File: packages/components/src/Enumeration/Items/Items.component.js [hunk 1]
```diff
@@ -69,8 +68,10 @@ function Items({ items, itemsProp, currentEdit, searchCriteria }) {
 				actions: itemsProp.actionsDefault,
 				onSelectItem: itemsProp.onSelectItem,
 			};
+			if (isSelectable) {
```
Comment (ncomont) [inline packages/components/src/Enumeration/Items/Items.component.js]: Could be `itemPropDefault.isSelectable = !!isSelectable;`

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://379.talend.surge.sh)

File: packages/components/src/Enumeration/Enumeration.component.js [hunk 2]
```diff
@@ -69,29 +66,63 @@ Enumeration.propTypes = {
...
+Enumeration.defaultProps = {
+	displayMode: DISPLAY_MODE_DEFAULT,
+	headerLabel: 'Values',
+	emptyLabel: 'This list is empty, click on + to add a value.',
+	toggleAllLabel: 'All',
+	items: [],
+};
+
+function ItemsEnumeration(props) {
+	// Default
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Enumeration.component.js]: I'm not sure those comments are required.

File: packages/components/src/Enumeration/Enumeration.component.js [hunk 3]
```diff
@@ -69,29 +66,63 @@ Enumeration.propTypes = {
...
+	const { items, itemsProp, emptyLabel } = props;
+	const { toggleAllChecked, toggleAllLabel, onToggleAll } = props;
+	// Edit
+	const { currentEdit } = props;
+	// Search
+	const { searchCriteria } = props;
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Enumeration.component.js]: All of those allocation can be avoided.

File: packages/components/src/Enumeration/Enumeration.component.js [hunk 4]
```diff
@@ -69,29 +66,63 @@ Enumeration.propTypes = {
...
+		toggleAllChecked,
+		toggleAllLabel,
+		onToggleAll,
+		emptyLabel,
+	};
+	return (
+		<Items
+			{...itemsProps}
+		/>
+	);
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Enumeration.component.js]: I don't think all of that props handling is actually required.
Most of the comments, allocations and object spread can be removed.

File: packages/components/src/Enumeration/Header/Header.defaultProps.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export default {
+	headerLabel: 'Values',
+};
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Header/Header.defaultProps.js]: default props on an external file ?

File: packages/components/src/Enumeration/Items/Item/Item.component.js [hunk 6]
```diff
@@ -86,30 +86,37 @@ function Item({ id, item, searchCriteria }) {
...
-			onClick={event => onSelectItem(item, event)}
-			className={itemLabelClasses()}
-			tooltip
-		/>);
+					<label htmlFor={itemId}>
+						<input
+							id={itemId}
+							type="checkbox"
+							onClick={event => onSelectItem(item, event)}
+							checked={!!item.isSelected}
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Items/Item/Item.component.js]: Not not ?

File: packages/components/src/Enumeration/Items/Item/ItemEdit.component.js [hunk 7]
```diff
@@ -130,25 +119,44 @@ class ItemEdit extends React.Component {
...
+				key={this.props.item.index}
+			>
+				<div className="checkbox-container">
+					<div className="checkbox">
+						<label htmlFor={itemId}>
+							<input
+								id={itemId}
+								type="checkbox"
+								onChange={event => this.props.item.itemsProp.onSelectItem(this.props.item, event)}
+								checked={!!this.props.item.isSelected}
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Items/Item/ItemEdit.component.js]: Not not ?

File: packages/components/src/Enumeration/Items/Items.component.js [hunk 8]
```diff
@@ -2,54 +2,136 @@ import React, { PropTypes } from 'react';
...
+		}
+		const computedIndex = this.hasToggleAll ? index - 1 : index;
+		const item = this.props.items[computedIndex];
+		const isEditMode = item && item.displayMode === DISPLAY_MODE_EDIT;
+		return this.props.itemsProp.getItemHeight(isEditMode);
+	}
+
+	getRowCount() {
+		const { items } = this.props;
+		return this.hasToggleAll ? items.length + 1 : items.length;
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Items/Items.component.js]: Please remove assignation and ternary.

File: packages/components/src/Enumeration/Items/Items.component.js [hunk 9]
```diff
@@ -2,54 +2,136 @@ import React, { PropTypes } from 'react';
...
+				this.props.itemsProp.onLoadData();
+			}
+		}, 500);
+	}
+
+	rowRenderer(props) {
+		const {
+			key,   // eslint-disable-line react/prop-types
+			index, // eslint-disable-line react/prop-types
+			style, // eslint-disable-line react/prop-types
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Items/Items.component.js]: Why disable the react-proptypes.

File: packages/components/src/Enumeration/Items/Items.component.js [hunk 10]
```diff
@@ -2,54 +2,136 @@ import React, { PropTypes } from 'react';
...
+		} = props;
+		const isToggle = this.hasToggleAll && index === 0;
```
Comment (talend-glorieux) [inline packages/components/src/Enumeration/Items/Items.component.js]: Unnecessary assignation

File: packages/components/stories/Enumeration.js [hunk 11]
```diff
@@ -125,15 +145,21 @@ selectedValuesProps.items = Array(50).fill('').map((item, index) => ({
 	isSelected: index % 2 === 0,
 }));
 
+if (selectedValuesProps.items.filter(item => !!item.isSelected).length > 1) {
+	const computedHeaderDefault = selectedValuesProps.headerDefault.slice();
+	computedHeaderDefault.unshift(deleteItemAction);
+	selectedValuesProps.headerDefault = computedHeaderDefault;
+}
```
Comment (talend-glorieux) [inline packages/components/stories/Enumeration.js]: That entire block of code is too complex. Please refactor it.

File: packages/forms/src/widgets/EnumerationWidget/EnumerationWidget.handlers.js [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+		if (this.timerSearch !== null) {
+			clearTimeout(this.timerSearch);
+		}
+		this.timerSearch = setTimeout(() => {
+			this.timerSearch = null;
+			if (this.callActionHandler(
+					ENUMERATION_SEARCH_ACTION,
+					value.value,
+					onSearchHandler.bind(this)
+				)) {
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/EnumerationWidget/EnumerationWidget.handlers.js]: An if statement calls a function with a binded aregument.
Please simplify.

File: packages/forms/src/widgets/EnumerationWidget/EnumerationWidget.handlers.js [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+		});
+		return;
+	}
+
+	if (this.callActionHandler(
+			ENUMERATION_ADD_ACTION,
+			value.value,
+			this.addSuccessHandler.bind(this),
+			this.addFailHandler.bind(this))
+	) {
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/EnumerationWidget/EnumerationWidget.handlers.js]: Same here. Simplify the if statement.

File: packages/forms/src/widgets/EnumerationWidget/utils/utils.js [hunk 14]
```diff
@@ -1,45 +1,27 @@
...
 export function manageCtrlKey(indexSelected, items) {
 	const itemsList = [...items];
 	const itemSelected = itemsList[indexSelected] && itemsList[indexSelected].isSelected;
-	if (itemSelected) {
-		itemsList[indexSelected].isSelected = false;
-	} else {
-		itemsList[indexSelected].isSelected = true;
-	}
+
+	itemsList[indexSelected].isSelected = !itemSelected;
```
Comment (talend-glorieux) [inline packages/forms/src/widgets/EnumerationWidget/utils/utils.js]: So isSelected is the opposite of itemSelected?

**Comment by jsomsanith** [comment]
close for inactivity

**Pull request closed without merging**